base/shell/explorer/taskswnd.cpp

```diff
@@ -678,7 +982,147 @@ class CTaskSwitchWnd :
         return m_ButtonCount;
     }
 
-    INT AddTaskItemButton(IN OUT PTASK_ITEM TaskItem)
+    BOOL CollapseTaskGroup(PTASK_GROUP TaskGroup) {
+        ASSERT(m_IsGroupingEnabled);
+
+        if (TaskGroup->IsCollapsed)
+            return FALSE;
+
+        /* Get the items associated to this group and delete the index */
+        PTASK_ITEM TaskItem, LastTaskItem = NULL;
+        WCHAR windowText[255];
+        TBBUTTON tbBtn = { 0 };
+        INT iIndex = INT_MAX;
+        HICON icon;
+
+        TaskItem = m_TaskItems;
+        LastTaskItem = TaskItem + m_TaskItemCount;
+
+        while (TaskItem != LastTaskItem)
+        {
+            if (TaskItem->Group == TaskGroup)
+            {
+                if (TaskItem->Index >= 0)
+                {
+                    if (TaskItem->Index < iIndex)
+                        iIndex = TaskItem->Index;
+                    DeleteTaskItemButton(TaskItem);
+                }
+            }
+
+            TaskItem++;
+        }
+
+        /* Open process to retrieve the filename of the executable */
+        WCHAR ExePath[MAX_PATH] = {};
+        if (GetProcessPath(TaskGroup->dwProcessId, ExePath, _countof(ExePath)))
+        {
+            if (GetVersionInfoString(ExePath, L"FileDescription", windowText, _countof(windowText)))
+                tbBtn.iString = (DWORD_PTR) windowText;
+
+            if (ExtractIconExW(ExePath, -1, NULL, &icon, 1) <= 0)
+                icon = static_cast<HICON>(LoadImageW(NULL, MAKEINTRESOURCEW(OIC_SAMPLE), IMAGE_ICON, 0, 0, LR_SHARED | LR_DEFAULTSIZE));
+        }
+
+        TaskGroup->IconIndex = ImageList_ReplaceIcon(m_ImageList, -1, icon);
```

Comment:
Note that as in the other `UpdateTaskGroupButton()` function, the `icon` variable is left being used uninitialized in case the `GetProcessPath` call fails.